jupyterhub/tests/test_roles.py

```diff
@@ -1046,3 +1046,71 @@ async def test_oauth_allowed_roles(app, create_temp_role):
     app_service = app.services[0]
     assert app_service['name'] == 'oas1'
     assert set(app_service['oauth_roles']) == set(allowed_roles)
+
+
+async def test_user_group_roles(app, create_temp_role):
+    user = add_user(app.db, app, name='jack')
+    another_user = add_user(app.db, app, name='jill')
+
+    group = orm.Group.find(app.db, name='A')
+    if not group:
+        group = orm.Group(name='A')
+        app.db.add(group)
+        app.db.commit()
+
+    if group not in user.groups:
+        user.groups.append(group)
+        app.db.commit()
+
+    if group not in another_user.groups:
+        another_user.groups.append(group)
+        app.db.commit()
+
+    group_role = orm.Role.find(app.db, 'student-a')
+    if not group_role:
+        create_temp_role(['read:groups!group=A'], 'student-a')
+        roles.grant_role(app.db, group, rolename='student-a')
+        group_role = orm.Role.find(app.db, 'student-a')
+
+    # repeat check to ensure group roles don't get added to the user more than once
```

Comment:
Group roles are not added to user's roles at all, they are only temporarily added in the `expand_roles_to_scopes` to return combined scope set for the user for further evaluation but the user never "physically" gets the group's role. In this case here the user has only default `user` role assigned upon creation in `user = add_user(app.db, app, name='jack')`.

Reply:
Resolved by @IvanaH8's code suggestion!